esrally/client/factory.py

```diff
@@ -238,3 +266,111 @@ def wait_for_rest_layer(es, max_attempts=40):
+    :return: True iff all provided key IDs were successfully deleted.
+    """
+    logger = logging.getLogger(__name__)
+
+    def raise_exception(failed_ids, cause=None):
+        msg = f"Could not delete API keys with the following IDs: {failed_ids}"
+        if cause is not None:
+            raise exceptions.RallyError(msg) from cause
+        else:
+            raise exceptions.RallyError(msg)
+
+    # Before ES 7.10, deleting API keys by ID had to be done individually.
+    # After ES 7.10, a list of API key IDs can be deleted in one request.
+    current_version = versions.Version.from_string(es.info()["version"]["number"])
+    minimum_version = versions.Version.from_string("7.10.0")
+
+    deleted = []
+    remaining = ids
+
+    for attempt in range(1, max_attempts + 1):
+        # pylint: disable=import-outside-toplevel
+        import elasticsearch
+
+        try:
+            if current_version >= minimum_version:
+                resp = es.security.invalidate_api_key({"ids": remaining})
+                deleted += resp["invalidated_api_keys"]
+                remaining = [i for i in ids if i not in deleted]
+                # Like bulk indexing requests, we can get an HTTP 200, but the
+                # response body could still contain an array of individual errors.
+                # So, we have to handle the case were some keys weren't deleted, but
+                # the request overall succeeded (i.e. we didn't encounter an exception)
+                if attempt < max_attempts:
+                    if resp["error_count"] > 0:
+                        logger.debug(
+                            "Got the following errors on attempt [%s] of [%s]: [%s]. Sleeping...",
+                            attempt,
+                            max_attempts,
+                            resp["error_details"],
+                        )
+                else:
+                    if remaining:
+                        logger.warning(
+                            "Got the following errors on final attempt to delete API keys: [%s]",
+                            resp["error_details"],
+                        )
+                        raise_exception(remaining)
+            else:
+                remaining = [i for i in ids if i not in deleted]
+                if attempt < max_attempts:
+                    for i in remaining:
+                        es.security.invalidate_api_key({"id": i})
+                        deleted.append(i)
+                else:
+                    if remaining:
+                        raise_exception(remaining)
+            return True
+
+        except elasticsearch.TransportError as e:
+            if attempt < max_attempts:
+                logger.debug("Got status code [%s] on attempt [%s] of [%s]. Sleeping...", e.status_code, attempt, max_attempts)
+                time.sleep(1)
+            else:
+                raise_exception(remaining, cause=e)
+        except Exception as e:
```

Comment:
Should we only catch `RallyError` here? I think for something else like say a `KeyError` there's no point in retrying.